File: 144_Binary_Tree_Preorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if (root == NULL) return vector<int>();
        stack<TreeNode*> s;
        vector<int> result;
        s.push(root);
        TreeNode* t;
        while (!s.empty()) {
            t = s.top();
            result.push_back(t->val);
            s.pop();
            if (t->right) s.push(t->right);
            if (t->left) s.push(t->left);
        }
        return result;
    }
};

File: 144_Binary_Tree_Preorder_Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        cur = root
        while cur is not None:
            if cur.left is None:
                res.append(cur.val)
                cur = cur.right
            else:
                pre = cur.left
                while pre.right is not None and pre.right != cur:
                    pre = pre.right
                if pre.right is None:
                    pre.right = cur
                    res.append(cur.val)
                    cur = cur.left
                else:
                    pre.right = None
                    cur = cur.right
        return res
